From 03532068fe017056cbd0481706ebc16eff12981b Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Tue, 5 Jul 2022 12:58:40 -0300
Subject: [PATCH 035/163] elf: Fix wrong fscanf usage on tst-pldd

The fix done b2cd93fce666fdc8c9a5c64af2741a8a6940ac99 does not really
work since macro strification does not expand the sizeof nor the
arithmetic operation.

Checked on x86_64-linux-gnu.

(cherry picked from commit c353689e49e72f3aafa1a9e68d4f7a4f33a79cbe)
---
 elf/tst-pldd.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/elf/tst-pldd.c b/elf/tst-pldd.c
index 8916ce5a2e..771c28a9c5 100644
--- a/elf/tst-pldd.c
+++ b/elf/tst-pldd.c
@@ -106,15 +106,16 @@ do_test (void)
      loader and libc.  */
   {
     pid_t pid;
-    char buffer[512];
-#define STRINPUT(size) "%" # size "s"
+#define BUFFERLEN 511
+    char buffer[BUFFERLEN + 1];
+#define STRINPUT(size)  XSTRINPUT(size)
+#define XSTRINPUT(size) "%" # size "s"
 
     FILE *out = fmemopen (pldd.out.buffer, pldd.out.length, "r");
     TEST_VERIFY (out != NULL);
 
     /* First line is in the form of <pid>: <full path of executable>  */
-    TEST_COMPARE (fscanf (out, "%u: " STRINPUT (sizeof (buffer) - 1), &pid,
-			  buffer), 2);
+    TEST_COMPARE (fscanf (out, "%u: " STRINPUT (BUFFERLEN), &pid, buffer), 2);
 
     TEST_COMPARE (pid, *target_pid_ptr);
     TEST_COMPARE (strcmp (basename (buffer), "tst-pldd"), 0);
-- 
2.37.2

